### PR TITLE
Generalized interfaces of algorithm interfaces for lambdas

### PR DESCRIPTION
**Pull request by fuchsto**
Addresses #274

### CODE REVIEW

**Review by devreal — CHANGES_REQUESTED** [review]
Please document the expected signature of the function. Also, members like `dash::generate` should be adopted as well (not sure where else support for lambdas is missing)

**Comment by devreal** [comment]
I noticed that `dash::generate` and `dash::[min|max]_element` are affected as well.

**Comment by fuchsto** [comment]
@devreal Good catch, will extend the PR to the affected algorithm interfaces.

**Comment by fuchsto** [comment]
Added documentation of template parameters (mostly function signatures), generalized function parameters to deduced template types.
I ran a `grep -R 'std::function' dash/include/dash/algorithm/*` which now comes back clean.

@devreal Should cover your comments.

File: dash/include/dash/algorithm/ForEach.h [hunk 1]
```diff
@@ -96,25 +58,27 @@ void for_each(
  * function on its local elements only. The index passed to the function is
  * a global index.
  *
- * \tparam      ElementType  Type of the elements in the sequence
- * \tparam      IndexType    Parameter type expected by function to
- *                           invoke, deduced from parameter \c func
+ * \tparam      ElementType            Type of the elements in the sequence
+ * \tparam      UnaryFunctionWithIndex Function to invoke for each element
+ *                                     in the specified range with signature
+ *                                     \c void (const ElementType &, index_t)
```
Comment (devreal) [inline dash/include/dash/algorithm/ForEach.h:68]: It does not have to be const, right?

File: dash/include/dash/algorithm/MinMax.h [hunk 2]
```diff
@@ -137,24 +140,26 @@ const ElementType * min_element(
...
  * \complexity  O(d) + O(nl), with \c d dimensions in the global iterators'
  *              pattern and \c nl local elements within the global range
  *
  * \ingroup     DashAlgorithms
  */
 template<
-  typename ElementType,
-  class    PatternType>
+  class ElementType,
+  class Compare = std::less<const ElementType &>,
```
Comment (devreal) [inline dash/include/dash/algorithm/MinMax.h]: Put `compare` (with default value) to the end of the the template parameter list?

File: dash/include/dash/algorithm/ForEach.h [hunk 3]
```diff
@@ -14,65 +14,27 @@ namespace dash {
...
+  class    PatternType,
+  class    UnaryFunction >
 void for_each(
   /// Iterator to the initial position in the sequence
   const GlobIter<ElementType, PatternType> & first,
   /// Iterator to the final position in the sequence
   const GlobIter<ElementType, PatternType> & last,
   /// Function to invoke on every index in the range
-  ::std::function<void(ElementType &)> & func)
+  UnaryFunction                              func)
```
Comment (devreal) [inline dash/include/dash/algorithm/ForEach.h:41]: `UnaryFunction` can be passed as `const &` (saves same cycles with `std::function`)

**Review by devreal — DISMISSED** [review]
Looks good, a few minor comments inline.

File: dash/include/dash/algorithm/ForEach.h (hunk 1 above)
Comment (fuchsto) [reply dash/include/dash/algorithm/ForEach.h:68]: Does not "have to", but it should:

<http://en.cppreference.com/w/cpp/algorithm/for_each>

The qualifier merely states that accepting const references to values is supported.

File: dash/include/dash/algorithm/ForEach.h (hunk 3 above)
Comment (fuchsto) [reply dash/include/dash/algorithm/ForEach.h:41]: Oh, no:
If `const Func &` is passed, the type `UnaryFunction` would resolve to `const Func &`.
When an lvalue function object is passed, the compiler will deduce `const T &`.
With `const UnaryFunction &`, non-const parameters would be explicitly forbidden.

Also see the signature of `std::for_each`:

<http://en.cppreference.com/w/cpp/algorithm/for_each>

File: dash/include/dash/algorithm/MinMax.h (hunk 2 above)
Comment (fuchsto) [reply dash/include/dash/algorithm/MinMax.h]: Definitely, fixed in my upcoming commit.

**Comment by fuchsto** [comment]
@devreal Updated documentation and fixed tparam order.

File: dash/include/dash/algorithm/ForEach.h (hunk 1 above)
Comment (fuchsto) [reply dash/include/dash/algorithm/ForEach.h:68]: Documentation has been extended and now also explicitly refers to `std::for_each`.

**Pull request merged**